assets/js-text-with-image.js

```diff
@@ -3,79 +3,168 @@
-    if (!elements.prevSection) return;
+    const elements = getElements(),
+          prevSection = elements.prevSection;
 
-    const prevChildren = [...elements.prevSection.children];
-    if (!prevChildren.length) return;
+    if (!prevSection) return;
 
-    elements.children.forEach(child => {
-      if (!child.classList.contains(config.classes.spacer)) return;
+    const prevChildren = prevSection.children;
+    if (!prevChildren || prevChildren.length === 0) return;
 
-      // Apply palette from each element in previous section
-      prevChildren.forEach(prevChild => {
-        applyPaletteToSpacer(child, prevChild);
-      })
-    })
+    const children = elements.section.children,
+          spacers = [],
+          prevWithPalettes = [];
+
+    // Use a single function for the detection work
+    const findElements = () => {
+      for (let i = 0; i < children.length; i++) {
+        if (hasSpacerClass(children[i])) {
+          spacers.push(children[i]);
+        }
+      }
+
+      if (spacers.length === 0) return false;
+
+      // Find all previous children with palettes
+      for (let i = 0; i < prevChildren.length; i++) {
+        const prevChild = prevChildren[i];
+        let hasPaletteClass = false;
+
+        // Check if this previous child has any palette class
+        for (let j = 0; j < config.class.palettes.length; j++) {
+          if (hasPalette(prevChild, config.class.palettes[j])) {
+            hasPaletteClass = true;
+            break;
+          }
+        }
+
+        if (hasPaletteClass) {
+          prevWithPalettes.push(prevChild);
+          break;
+        }
+      }
+
+      // If no previous children with palettes, signal early exit
+      return prevWithPalettes.length > 0;
+    }
+
+    if (!findElements()) return;
+
+    // Use the first previous child with a palette for simplicity
+    const sourcePalette = prevWithPalettes[0];
+    const applyPalettes = () => {
+      for (let i = 0; i < spacers.length; i++) {
+        applyPaletteToSpacer(spacers[i], sourcePalette);
+      }
+    }
+
+    // Use batch DOM operations for better performance
+    $.batchDOM(applyPalettes);
   }
 
-  const initialize = () => setSpacerColors();
+  const initialize = () => {
+    setSpacerColors();
+    return true;
+  }
 
   return {
-    initialize
+    initialize,
+    updateColors: setSpacerColors // Allow manual updates if needed
   }
 }
 
 const initTextWithImageSpacer = (selector = '.text-image') => {
-  const sections = document.querySelectorAll(selector);
+  const instances = [],
+        sections = document.querySelectorAll(selector);
   if (!sections.length) return;
 
-  sections.forEach(section => {
-    const spacer = handleTextWithImageSpacer(section);
-    if (spacer) spacer.initialize();
-  })
+  const processSections = () => {
+    sections.forEach(section => {
+      const spacer = handleTextWithImageSpacer(section);
+      if (spacer && spacer.initialize()) instances.push(spacer);
+    })
```

Comment:
Here maybe we could use `filter`?

Also, it seems `initialize` always return true. We can just check for the spacer? Something like:

```js
const spacer = handleTextWithImageSpacer(section);
if (!spacer) return;

spacer.initialize();
instances.push(spacer);
```